src/cfg/domtree.h

```diff
@@ -0,0 +1,157 @@
+//
+// The BasicBlock type is assumed to have a ".in" property which declares a
+// vector of pointers to the incoming blocks, that is, the predecessors.
+template<typename BasicBlock> struct DomTree {
+  std::vector<Index> parents;
+
+  DomTree(std::vector<std::unique_ptr<BasicBlock>>& blocks);
+};
+
+template<typename BasicBlock>
+DomTree<BasicBlock>::DomTree(std::vector<std::unique_ptr<BasicBlock>>& blocks) {
+  // Compute the dominator tree using the "engineered algorithm" in [1]. Minor
+  // differences in notation from the source include:
+  //
+  //  * doms => parents. The final structure we emit is the vector of parents in
+  //    the dominator tree, and that is our public interface, so name it
+  //    explicitly.
+  //  * Indexes are reversed. The paper uses postorder indexes, i.e., the entry
+  //    node has the highest index, while we have a natural indexing since we
+  //    traverse in reverse postorder anyhow (see cfg-traversal.h), that, is,
+  //    the entry has the lowest index.
+  //  * finger1, finger2 => left, right.
+  //
+  // Otherwise this is basically a direct implementation. You can ignore the
+  // comments here if you are already familiar with the algorithm.
+  //
+  // [1] Cooper, Keith D.; Harvey, Timothy J; Kennedy, Ken (2001). "A Simple,
+  //       Fast Dominance Algorithm" (PDF).
+  //       http://www.hipersoft.rice.edu/grads/publications/dom14.pdf
+
+  // If there are no blocks, we have nothing to do.
+  Index numBlocks = blocks.size();
+  if (numBlocks == 0) {
+    return;
+  }
+
+  // Map basic blocks to their indices.
+  std::unordered_map<BasicBlock*, Index> blockIndices;
+  for (Index i = 0; i < numBlocks; i++) {
+    blockIndices[blocks[i].get()] = i;
+  }
+
+  // Use a nonsense value to indicate what has yet to be initialized.
+  const Index nonsense = -1;
+
+  // Initialize the parents array. The entry starts with its own index, which is
+  // used as a guard value in effect (we will never process it, and we will fix
+  // up this value at the very end). All other nodes start with a nonsense value
+  // that indicates they have yet to be processed.
+  parents.resize(numBlocks);
+  parents[0] = 0;
+  for (Index i = 1; i < numBlocks; i++) {
+    parents[i] = nonsense;
+  }
+
+  // Loop over the (non-entry) blocks in reverse postorder while there are
+  // changes still happening.
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (Index index = 1; index < numBlocks; index++) {
+      // Loop over the predecessors. Our new parent is basically the
+      // intersection of all of theirs: our immediate dominator must precede all
+      // of them.
+      auto& preds = blocks[index]->in;
+      Index newParent = nonsense;
+      for (auto* pred : preds) {
+        auto predIndex = blockIndices[pred];
+        if (parents[predIndex] == nonsense) {
+          // This pred has yet to be processed; we'll get to it in a later
+          // cycle.
+          continue;
+        }
```

Comment:
This only happens for loop backedges, which should be dominated by the current block (loop header) in reducible CFGs, and thus don't affect the idom computation. I guess this can also happen for unreachable blocks but I'm not sure that can be created from a Binaryen AST? Anyway they don't affect idom computation either. Then do we need to "get to it in a later cycle"?

(I think, if we don't consider unreachable blocks, we can even simplify the loop to
```cpp
if (predIndex > index) {
  continue;
}
```
)

Also, do we need the outer `while (changed)` in that sense? If we don't consider irreducible CFGs, it looks the process can be completed with a single traversal of `blocks` without the `while` loop. It wouldn't be able to handle the irreducible CFG in your last example, but we don't need to do that for Binaryen anyway. If we have this `while (changed)`, every `DomTree` construction will traverse the `blocks` twice instead of once, even for reducible CFGs, because `changed` will be set to true for every CFG in its first traversal.

Reply:
Good points! Yes, we may as well simplify things given reducibility. I was focused on implementing the algorithm as is, but we really don't need that.